model-optimizer/extensions/front/tf/NonConstBeginStridedSliceReplacement.py

```diff
@@ -0,0 +1,136 @@
+    The transformation attempts to match to following sub-graph:
+
+    Input ----> StridedSlice(begin_mask, end_mask, and shrink_mask where only element for axis equal to 1) --> OTHER OPS
+                   /\                  /\                  /\
+                   |                   |                   |
+      ---------> Pack(Begin)          Pack(End)         Const(Step) = (1,..,1)
+      |               /\                /\   /\
+      |               |                 |    |
+      |               |                 |   Const(All others)
+      |        Const(All others)        |
+    Index ---------------------------> Add
+                                        /\
+    Const(SliceSize)=1------------------|
+
+    And the original sub-graph is transformed as follows:
+
+    Input --------> Gather(axis) ---> Squeeze(axis) ---> OTHER OPS
+                       /\
+                       |
+    Index -----> Unsqueeze(axis=1)
+
+    """
+    enabled = True
+
+    @staticmethod
+    def pattern(**kwargs):
+        return dict(
+            nodes=[('begin', dict(op='Pack')),
+                   ('end', dict(op='Pack')),
+                   ('step', dict(op='Const')),
+                   ('strided_slice', dict(op='StridedSlice')),
+                   ],
+            edges=[('begin', 'strided_slice', {'in': 1}),
+                   ('end', 'strided_slice', {'in': 2}),
+                   ('step', 'strided_slice', {'in': 3})])
+
+    def replace_sub_graph(self, graph: Graph, match: dict):
+        strided_slice_node = match['strided_slice']
+        begin_node = match['begin']
+        end_node = match['end']
+        step_node = match['step']
+
+        # retrieve attribute values
+        begin_mask = strided_slice_node.soft_get('begin_mask')
+        end_mask = strided_slice_node.soft_get('end_mask')
+        shrink_mask = strided_slice_node.soft_get('shrink_axis_mask', int64_array([0]))
+
+        # check applicability of this transformation to the given sub-graph:
+        # 1. check that slicing is performed along just one axis
+        if np.sum(begin_mask) != 1 or np.sum(end_mask) != 1 or np.sum(shrink_mask) != 1:
+            return
+        # 2. check that shrink axis is equal to slicing axis
+        if not np.array_equal(np.argwhere(begin_mask == 1), np.argwhere(end_mask == 1)) or \
+                not np.array_equal(np.argwhere(begin_mask == 1), np.argwhere(shrink_mask == 1)):
+            return
+        sliced_axis = np.argwhere(begin_mask == 1)[0][0]
+        # 3. check constant nodes for begin and end correspond to non-slicing axes
+        for idx_port, in_port in begin_node.in_ports().items():
+            if idx_port != sliced_axis and in_port.get_source().node.soft_get('type') != 'Const' or \
+                    idx_port == sliced_axis and in_port.get_source().node.soft_get('type') == 'Const':
+                return
```

Comment:
Can be simplified to:

```python
if (idx_port == sliced_axis) == (in_port.get_source().node.soft_get('type') == 'Const'):
    return
```

Reply:
It looks tricky to me. I like more explicit logic syntax.